docs/graph_engine/custom_decoder.rst

```diff
@@ -85,11 +85,11 @@ index 0 and 2 but skip index 1.
     ...         # Args can be added to init and added before passing to MultiWorkersConverter
     ...         pass
     ...
-    ...     def decode(self, line: dict) -> Iterator[Tuple[int, int, int, float, list]]:
+    ...     def decode(self, line: dict) -> Iterator[Tuple[int, int, int, float, int, int, list]]:
     ...         # For a node we will yield: (node_id, -1, node_type, node_weight, [feature_0, ...])
     ...         # For an edge: (src_id, dst_id, edge_type, edge_weight, [feature_0, ...])
     ...         # We can return multiple items per line, but the order of the file, described above, must be maintained.
-    ...         yield int(line["src"]), int(line["dst"]), int(line["type"]), float(line["weight"]), [np.array(line["float_feature"], dtype=np.float32), (np.array(line["sparse_int16_coords"], dtype=np.int64), np.array(line["sparse_int16_feature"], dtype=np.int16))]
+    ...         yield int(line["src"]), int(line["dst"]), int(line["type"]), float(line["weight"]), None, None, [np.array(line["float_feature"], dtype=np.float32), (np.array(line["sparse_int16_coords"], dtype=np.int64), np.array(line["sparse_int16_feature"], dtype=np.int16))]
```

Comment:
edge list needs to support temporal features as well.

Reply:
it supports it.

Reply:
What? Just make if user puts in a 3 tuple (ndarray, int, int) then that is a temporal feature

Reply:
There are no ndarrays in text format. You are mixing decoder interface and parsing.

Reply:
I see,
features[dense]: dtype,length,v1,v2,...,dtype2,length2,v1,v2,...
features[sparse]: dtype,values_length/coords_dim,c1,c2,...,v1,v2,...
features[temporal_dense]: dtype,length/created_at/deleted_at,v1,v2,...
features[temporal_sparse]: dtype,values_length/coords_dim/created_at/deleted_at,c1,c2,...,v1,v2,...

Reply:
what is this?

Reply:
This is a way to update the text format to include temporal features without conflicts

Reply:
features[temporal_dense]: dtype,length/created_at/deleted_at,v1,v2,...
E.g. 
float32,3/0/100,0,1,2
= data=np.array([0, 1, 2], dtype=float32), created_at=0, deleted_at=100

Reply:
Feature values that are created and removed multiple times. Example from doc:
```json
"sparse_int32_feature": {
	"0": [
		{"coordinates": [[5, 13], [7, 25]], "values": [-1, 1024], "removed_at": 21},
		{"coordinates": [[4, 2], [1, 3]], "values": [2, 4], "created_at": 21}
	],
	"1": {"coordinates": [13, 42], "values": [1, 1]}
}
```

Reply:
Is there something blocking us from just having multiple entries per feature id in the edge list format?

Reply:
[Features will be given indexes starting at 0 and indexes can be skipped with 0 length vectors.](https://github.com/microsoft/DeepGNN/blob/0abd02f3c11ea8e051b6d224f8a33954b7d3d90c/src/python/deepgnn/graph_engine/snark/decoders.py#L64)

Reply:
features[temporal_dense]: dtype,length/created_at/deleted_at,v1,v2,...length2/created_at2/deleted_at2,v1,v2,...
Instead of having next dtype just have another tuple and it will recognize as a second temporal value.

Reply:
There are no tuples in text. Please write explicit contents of a graph with the node feature from above and state why does it have to be added in this PR.

Reply:
> Parsing needs to be added to this PR since EdgeList is the fastest format for conversion that we support and it needs to be fully featured.

This is a circular logic. `A` has to be added, because we have to have `A`. I am also pretty sure [the avro decoder in docs](https://github.com/microsoft/DeepGNN/blob/main/docs/graph_engine/custom_decoder.rst) is faster than text parsing with edge lists, because the former is binary.

> int32,2/2/0/21,5,13,7,25,-1,1024,2/2/21/100,4,2,1,3,2,4,int32,2/1,13,42,1,1

The proposed line doesn't match json example, there are no `created_at` value in the first feature(with values -1, 1024) and `removed_at` in the second(with values 2, 4). Supporting default values in positional encoding is a hard problem. Because the next item,

> int32,2/1,13,42,1,1

is indistinguishable from a dense feature, with `created_at 2`, `removed_at` time 1 and 4 feature values [13,42,1,1].

tl;dr In the current version edge list format is ambiguous for temporal features, hence we can't use it. If anyone requests it, we can design it properly and add in the future, but I don't see a reason to add it right now.